### PR TITLE
animation for transform

### PR DESCRIPTION
**Pull request by Jason0214**
### finish animation for node and skeleton transform
    functional it works, but still have several details to discuss

1. **the animation export format** 
in my implementation, if a node has animation data while there are no animation data in all its ancestors, then it will exported an AnimationPlayer as its sibling, besides all the animation data in its descendants all goes into that AnimationPlayer. I supposed it is what user wanted in most case.
2. **the code structure** 
with animation, now there are two traversal pass for the scene tree, the first pass is applied to Blender scene tree, dealing with skeleton, mesh and etc, also if there is animation data found in Blender node, then assign the Blender node to the corresponding Godot scene node's ```anim_data```. Then the second pass is applied to the Godot scene tree, dealing with all the found ```anim_data```. 
There are two problems, first the ```anim_data``` (works as a temporary variable) is inappropriate attribute of```NodeTemplate```, may be I should inherit ```NodeTemplate```?  besides, the ```animation.py``` is not fall into the category of converters,
3. **fix_matrix function**
```fix_matrix```(used to convert a Blender matrix to Godot matrix) is a local function inside ```mat_to_string()```, however it need to be used in ```animation.py```, I am thinking about extract it into a global accessible position, but not sure where to put it
4. **optimization in keyframe**
since Godot doesn't support Bezier interpolation, I exported animation by treat every frame as keyframe(maybe it's called baking?), it is definitely need to do some optimization to remove some not important keyframe, it is not implemented yet. I thought the collada importer of godot engine may have some code about it, but I could not find it (anyone give me a hint?)
5. pylint warning for too many local variables, too many nested blocks, too many branches. I don't know how to solve it .. :P 
6. can we have a interface to add a internal_resource without providing a hashable object, it is hard to  have a hash for animations

### CODE REVIEW

File: io_scene_godot/converters/animation.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+    for path, obj in anim_objects.items():
+        orig_nla_status.append(obj.animation_data.use_nla)
+        obj.animation_data.use_nla = False
+
+    anim_data_id = export_animation_data(escn_file, 'default',
+                                         anim_objects, anim_root_path)
+    anim_player_node['anims/default'] = "SubResource({})".format(anim_data_id)
+
+    # save back status
```
Comment (sdfgeoff) [inline io_scene_godot/converters/animation.py]: I think blender stores and undoes any changes that the export script performs, so you don't need to worry about making changes to the blend.

File: io_scene_godot/converters/animation.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    def __init__(self, timestamp, data):
+        self.timestamp = timestamp
+        self.data = data
+
+    @staticmethod
+    def identical(key_type, key1, key2):
+        """Check if two keyframes has the same
+        conversion data"""
+        if key_type == 'transform':
+            return key1.data == key2.data
```
Comment (sdfgeoff) [inline io_scene_godot/converters/animation.py]: Is this general enough to get rid of the check for being type 'transform'?

**Comment by sdfgeoff** [comment]
Disclaimer: animation is not my strong point, so some of the things here are a bit of my guesswork. I know only a little of blender's animation toolbox, and virtually none of Godot's. As a result, it may be good to get some input from @reduz and if there are knowledgeable blender animators around, it would be good to get their input as well.

1. If I understand what you're saying, then so long as it isn't the child of something with an animation, then it will get it's own AnimationPlayer. (So if I have a level with multiple doors, then if the doors aren't parented to each other, each will get a unique AnimationPlayer. But if I have a sword parented to an animated character, then the sword will use the characters AnimationPlayer.
That sounds sane to me.

2. Hmm, is animation a converter? It doesn't convert a blender object to a godot node, but it does convert a blender concept into a godot concept. I think leave it in the `converters` folder for now. Animations do not need to be in the BLENDER_TYPE_TO_EXPORTER dict, just as the convert material function isn't. The BLENDER_TYPE_TO_EXPORTER is intended for direct node-to-node conversion on base blender types.
The main issue is that the exporter was structured for converting blender objects into nodes and blender data into resources, and in Blender, an animation is a data-like-thing, but in Godot an animation is a Node. As a result, the architecture of the exporter doesn't really fit properly with animations. So I suspect that any solution that works properly will be slightly "hacky". Particularly because we should consider if we can make it easy to add animatable properties in the future (eg colors, materials) as well as just transforms.

Perhaps animations.py could expose two functions: `get_animation_player(escn_file, godot_node)` function, and `export_animation(animation_player, godot_node, godot_param_name, blender_fcurve)` which writes the animation data into the animation player. (see bpy.types.FCurve) This would potentially:

 - remove the need for a second iteration
 - allow, for example, the light node to export the light animations without putting per-type information inside the animation converter
 - allow exporting animations that aren't currently being displayed (eg different poses for a character) by passing in different f-curves. (eg via iterating through bpy.data.actions, or [this thing here](https://blender.stackexchange.com/questions/78643/is-there-a-way-to-find-all-actions-an-object-is-involved-with))


3. You're already importing structures, and the fix_matrix function is inside structures, so moving fix_matrix to be at module-level inside `structures.py` should allow access from anywhere.

4. I notice that Godot does have the "cubic" option. I couldn't get it to have any noticable affect in my build, but in theory this should be similar to beziers (a bezier can be converted into a cubic). It may be worth asking @reduz about this.
cubic

File: io_scene_godot/converters/animation.py (hunk 1 above)
Comment (Jason0214) [reply io_scene_godot/converters/animation.py]: ok. I will remove it.

File: io_scene_godot/converters/animation.py (hunk 2 above)
Comment (Jason0214) [reply io_scene_godot/converters/animation.py]: I keep this because I plan to do some fuzzy comparison between adjacent frames to reduce the keyframe number

**Comment by sdfgeoff** [comment]
If you pass in bpy.types.FCurve, then you can extract all the data using the [evaluate](https://docs.blender.org/api/2.79/bpy.types.FCurve.html#bpy.types.FCurve.evaluate) function. This means you don't have to increment the scene.

**Comment by Jason0214** [comment]
reduz suggested to bake every frame in exporting animation. And when the escn file being imported into Godot, the animation would be optimized. Then I think this PR is ready to review.

**Comment by Jason0214** [comment]
woops, still have two bugs in this PR
1. the fcurve value is relative to the world not its parent 
2. the eular rotation of bone not converted to quaternion

**Comment by Jason0214** [comment]
Change back to the baking frame approach, since the convert the fcurve data from world space to parent space would modify the bezier curve of animation and lose data.  Now it's ready to review.

File: io_scene_godot/structures.py [hunk 3]
```diff
@@ -59,12 +61,14 @@ def get_internal_resource(self, hashable):
 
     def add_internal_resource(self, item, hashable):
         """See comment on external resources. It's the same"""
-        if self.get_internal_resource(hashable) is not None:
+        if (hashable is not None and
```
Comment (sdfgeoff) [inline io_scene_godot/structures.py]: So I guess you wanted to be able to pass in hashable=None to be able to force it to create a new resource. What is the use case of this? The idea is that every time you can pass in the blender _something_ that is being stored.

File: io_scene_godot/export_godot.py [hunk 4]
```diff
@@ -129,6 +133,14 @@ def export_scene(self):
             if obj in self.valid_nodes and obj.parent is None:
                 self.export_node(obj, root_gd_node)
 
+        # second pass to export animation
```
Comment (sdfgeoff) [inline io_scene_godot/export_godot.py]: I'm very keen to do the exports in a single "pass" because otherwise you may have issues with (for example) an object's animations being exported without the object itself. 
I think I'm missing the reason why this is necessary, so next time you're on IRC, I think we should discuss this.

File: io_scene_godot/converters/animation.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+    frame_len = 1 / bpy.context.scene.render.fps
+    frame_start = int(frame_start)
+    frame_end = int(frame_end)
+
+    # export the animation is done by baking frame because
+    # blender keyframe transform is relative to world space
+    # while godot transform is relative to parent space
+    for frame in range(frame_start, frame_end+1):
+        bpy.context.scene.frame_set(frame)
```
Comment (sdfgeoff) [inline io_scene_godot/converters/animation.py]: Stepping through the frames of the scene has several limitations, as it forces all of the animations to be active in the scene to be exported. In blender it is relatively standard to use an action library (eg to differentiate between a walk cycle and a run cycle). This concept is present in Godot as well (multiple animations in the same animationplayer) so even if this isn't supported now, it would be good to ensure the exporter can easily support it in the future.

As a result I would rather have some sort of function operating on the F-curves - even though it would have to evaluate them every frame.

Comment (Jason0214) [reply io_scene_godot/converters/animation.py]: If we deal with F-curves, it means we need to convert the F-curve keyframe value from world space to parent space, the conversion would guarantee the keyframe perfectly exported, however, as the value of F-curve changed, the interpolation would be changed, thus the whole animation (the non-keyframe frames) would be different.

If I do not mistake it, the approach to create something like an animation library in blender is to use the nla_stack, and nla_stack can be accessed through object->animation_data->[nla_track]->[nla_strip]->action, it can be easily exported

File: io_scene_godot/structures.py (hunk 3 above)
Comment (sdfgeoff) [reply io_scene_godot/structures.py]: You can pass in a reference to a blender object, so you could pass in a bpy.types.Action as the hashable (yes, it's mutable, but it's a C-based structure, so python treats it's memory address as the hash)

Comment (sdfgeoff) [reply io_scene_godot/structures.py]: In discussion via IRC this was resolved in favor of the existing code because while an f-curve or action may be unique, there is nothing that represents Godot's AnimationPlayer node.

Comment (Jason0214) [reply io_scene_godot/structures.py]: If you agree I still want to provide no hash for an animation resource. Use a bpy.types.action as a hash object is hacky (it does not one-one mapping to the animation resource) and also no good (animation is exporting in single traversal of blender tree, it never processes an action more that one time, so use a action would not catch any duplication)

**Comment by Jason0214** [comment]
request a review @sdfgeoff
there is a place I tagged FIXME, it's about the animation will lose negative scale, I am no idea what problem it can cause.

**Comment by sdfgeoff** [comment]
IRC Conversation copied here for storage

```
sdfgeoff | So my only thought is that the function "export_animation_track" should perhaps be
         | split up.
sdfgeoff | The reliance on TRANSFORM_ATTRIBUTES also seems like it could be a bit better
 JasonLu | so if it is to split "export_animation_track", I think then extract the fcurves ->
         | transform_frames_map out
      <snip>
sdfgeoff | Re: reliance: If the user has animated the color of an object, the energy of a lamp
         | etc. etc. then using TRANSFORM_ATTRIBUTES here is a bit limiting
sdfgeoff | If a developer in the future wants to add support for these they'll have to edit the
         | export_animation_track function to allow each new type of animation.
 JasonLu | exactly
sdfgeoff | I think that they shouldn't have to.
sdfgeoff | Currently the export_animation_track is all about exporting the transformation tracks
 JasonLu | emmm, then we should have a separate function to convert fcurves to transform track
 JasonLu | and when others want to support animation about color, they add a new function
sdfgeoff | I don't think they should have to add a new function. Surely a call to an existing
         | function should be adequate.
sdfgeoff | Something like create_track(node, fcurve, animation_name)
 JasonLu | the problem is that he value animation is quite different from transform animation, for
         | example it does not need to convert to matrix and convert back
sdfgeoff | create_track(node, fcurve, animation_name, function_to_run_on_every_keyframe_value) ?
 JasonLu | how about keeping the PR open, and I continue to finish the value animation,
 JasonLu | I first to do normally and  when complete we can discuss how to optimize it
sdfgeoff | Yup, that sounds good to me
sdfgeoff | My only thought is that it must be as easy as possible to add new animated properties -
         | ideally without having to add or change the code in converters/animation.py
 JasonLu | emm, I will consider it,  I suppose it would be possible for different kinds of value
         | transform (like color, shape key etc), but the transform track maybe can not be put
         | into that category
 JasonLu | ** value animation track **
sdfgeoff | In the end, I'm more concerned about functionality, so even if the transform track is a
         | special case, that should be fine.
```

**Comment by Jason0214** [comment]
have bug in fcurves of `posebone.matrix_basis`. 
some base knowledge
1. In Blender `bone.matrix_local  * posebone.matrix_basis` is the bone pose in armature space. 
2. In Godot `bone.rest * bone.pose` is the bone pose in parent bone space

I directly export `posebone.matrix_basis` to `bone.pose`, it works incorrect, maybe I made a mistake in 1 or 2.  

and another issue, blender pose bone has a weird location which i don't know what it is

File: io_scene_godot/structures.py [hunk 6]
```diff
@@ -283,30 +283,31 @@ def to_string(self):
...
-def mat4_to_string(mtx):
-    """Converts a matrix to a "Transform" string that can be parsed by Godot"""
-    def fix_matrix(mtx):
-        """ Shuffles a matrix to change from y-up to z-up"""
-        # Todo: can this be replaced my a matrix multiplcation?
-        trans = mathutils.Matrix(mtx)
-        up_axis = 2
+def fix_matrix(mtx):
+    """ Shuffles a matrix to change from y-up to z-up"""
+    # Todo: can this be replaced my a matrix multiplcation?
```
Comment (sdfgeoff) [inline io_scene_godot/structures.py]: I would make this my own commit except that I know you've made this edit here. But fix_matrix can be replaced by:

```python
def fix_matrix(mtx):
    """ Shuffles a matrix to change from y-up to z-up"""
    new_mat = mtx * mathutils.Matrix([
        [1, 0, 0, 0],
        [0, 0, -1, 0],
        [0, 1, 0, 0],
        [0, 0, 0, 1],
    ])
    new_mat[1], new_mat[2] = new_mat[2].copy(), -new_mat[1].copy()
    return new_mat
```

Comment (sdfgeoff) [reply io_scene_godot/structures.py]: Yeah, but in master branch it's in a different place, and when this gets merged the change may go unnoticed.

Comment (Jason0214) [reply io_scene_godot/structures.py]: then leave it after this PR merged, it is not urgent

**Comment by Jason0214** [comment]
@sdfgeoff  ready to have a review,  I add an option to create separate AnimationPlayer for each animated node (suggested by reduz).
